core/src/test/scala/unit/kafka/server/AbstractFetcherThreadTest.scala

```diff
@@ -123,4 +124,77 @@ class AbstractFetcherThreadTest {
       new DummyFetchRequest(partitionMap.map { case (k, v) => (k, v.offset) }.toMap)
   }
 
+
+  @Test
+  def testFetchRequestCorruptedMessageException() {
+    val partition = new TopicPartition("topic", 0)
+    val fetcherThread = new TestFetcherThread("test", "client", new BrokerEndPoint(0, "localhost", 9092))
+
+    fetcherThread.start()
+
+    // add one partition for fetching
+    fetcherThread.addPartitions(Map(partition -> 0L))
+
+    // wait until fetcherThread finishes the work
+    TestUtils.waitUntilTrue(() => (fetcherThread.fetchCount > 3),
+      "Failed waiting for fetcherThread tp finish the work")
+
+    fetcherThread.shutdown()
+
+    assertTrue(fetcherThread.logEndOffset == 2);
+  }
+
+  class TestFetcherThread(name: String,
+                          clientId: String,
+                          sourceBroker: BrokerEndPoint)
+    extends DummyFetcherThread(name, clientId, sourceBroker) {
+
+    var logEndOffset = 0L
+    var fetchCount = 0
+    val normalPartitionDataSet = List(new NormalPartitionData(Seq(0)), new NormalPartitionData(Seq(1)))
+
+    override def processPartitionData(topicAndPartition: TopicPartition,
+                                      fetchOffset: Long,
+                                      partitionData: PartitionData): Unit = {
+      if (fetchOffset != logEndOffset)
+        throw new RuntimeException(
+          "Offset mismatch for partition %s: fetched offset = %d, log end offset = %d."
+            .format(topicAndPartition, fetchOffset, logEndOffset))
+
+      val messages = partitionData.toByteBufferMessageSet
+      for (messageAndOffset <- messages.shallowIterator) {
+        val m = messageAndOffset.message
+        m.ensureValid()
+        logEndOffset = messageAndOffset.nextOffset
+      }
+    }
+
+    override protected def fetch(fetchRequest: DummyFetchRequest): Seq[(TopicPartition, DummyPartitionData)] = {
+      if (fetchCount == 0) {
+        fetchCount += 1
+        fetchRequest.offsets.mapValues(_ => new CorruptedPartitionData).toSeq
+      } else {
+        fetchCount += 1
+        fetchRequest.offsets.map {
+          case (k, v) => (k, normalPartitionDataSet(v.toInt))
+        }.toSeq
+      }
+    }
+  }
+
+  class CorruptedPartitionData extends DummyPartitionData {
+    override def toByteBufferMessageSet: ByteBufferMessageSet = {
+      val corruptedMessage = new Message("hello".getBytes)
+      val badChecksum: Int = (corruptedMessage.checksum + 1 % Int.MaxValue).toInt
```

Comment:
Nit: no need for `: Int` here since the type is obvious (we have the `toInt` call).